lib/Sema/CSDiag.cpp

```diff
@@ -4555,6 +4555,105 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
 }
 
 
+static bool diagnoseImplicitSelfErrors(Expr *fnExpr, Expr *argExpr,
+                                       bool hasTrailingClosure,
+                                       ArrayRef<Identifier> argLabels,
+                                       ConstraintSystem *CS) {
+  auto &TC = CS->TC;
+  // Call expression is formed as 'foo.bar' where 'foo' might be an
+  // implicit "Self" reference, such use wouldn't provide good diagnostics
+  // for situations where instance members have equal names to functions in
+  // Swift Standard Library e.g. min/max.
+  auto UDE = dyn_cast<UnresolvedDotExpr>(fnExpr);
+  if (!UDE)
+    return false;
+
+  auto baseExpr = dyn_cast<DeclRefExpr>(UDE->getBase());
+  if (!baseExpr)
+    return false;
+
+  auto baseDecl = baseExpr->getDecl();
+  if (!baseExpr->isImplicit() || baseDecl->getName() != TC.Context.Id_self)
+    return false;
+
+  // Our base expression is an implicit 'self.' reference e.g.
+  //
+  // extension Sequence {
+  //   func test() -> Int {
+  //     return max(1, 2)
+  //   }
+  // }
+  //
+  // In this example the Sequence class already has two methods named 'max'
+  // none of which accept two arguments, but there is a function in
+  // Swift Standard Library called 'max' which does accept two arguments,
+  // so user might have called that by mistake without realizing that
+  // compiler would add implicit 'self.' prefix to the call of 'max'.
+
+  // Get outermost context relative to this expression.
+  auto baseCtx = baseDecl->getDeclContext();
+  auto moduleCtx = baseCtx->getModuleScopeContext();
+  auto result = TC.lookupUnqualified(moduleCtx, UDE->getName(), UDE->getLoc());
```

Comment:
Maybe a silly question, but are we guaranteed that the results of this are exactly the same as the regular lookup that occurs (outside of attempting to diagnose this error)? I'm just wondering whether it's possible that this fix-it code would produce a qualified call that doesn't actually reference an accessible function.